lib/matplotlib/backends/backend_pdf.py

```diff
@@ -1256,73 +1261,94 @@ def imageObject(self, image):
         self.images[image] = (name, ob)
         return name
 
-    def _rgb(self, im):
-        h, w, s = im.as_rgba_str()
+    def _unpack(self, im):
+        """Unpack the image object im into height, width, data, alpha,
+        where data and alpha are HxWx3 (RGB) or HxWx1 (grayscale or alpha)
+        arrays, except alpha is None if the image is fully opaque."""
 
+        h, w, s = im.as_rgba_str()
         rgba = np.fromstring(s, np.uint8)
         rgba.shape = (h, w, 4)
         rgba = rgba[::-1]
-        rgb = rgba[:, :, :3].tostring()
-        a = rgba[:, :, 3]
-        if np.all(a == 255):
+        rgb = rgba[:, :, :3]
+        alpha = rgba[:, :, 3][..., None]
+        if np.all(alpha == 255):
             alpha = None
         else:
-            alpha = a.tostring()
-        return h, w, rgb, alpha
-
-    def _gray(self, im, rc=0.3, gc=0.59, bc=0.11):
-        rgbat = im.as_rgba_str()
-        rgba = np.fromstring(rgbat[2], np.uint8)
-        rgba.shape = (rgbat[0], rgbat[1], 4)
-        rgba = rgba[::-1]
-        rgba_f = rgba.astype(np.float32)
-        r = rgba_f[:, :, 0]
-        g = rgba_f[:, :, 1]
-        b = rgba_f[:, :, 2]
-        a = rgba[:, :, 3]
-        if np.all(a == 255):
-            alpha = None
+            alpha = np.array(alpha, order='C')
+        if im.is_grayscale:
+            r, g, b = rgb.astype(np.float32).transpose(2, 0, 1)
```

Comment:
On 09 Jul 2015, at 17:16, Benjamin Root <notifications@github.com> wrote:
> 
> > + if im.is_grayscale:
> 
> > +            r, g, b = rgb.astype(np.float32).transpose(2, 0, 1)
> just as a sanity check, when did this form of transpose() become available in numpy? It might have always been there, but I want to double-check.
> 

Not sure when exactly, but it’s documented in the 2006 copy of numpybook.pdf, which mentions version 1.0.2.dev3478. We support numpy 1.6 and up. One of the Travis builders uses Python 2.6 with numpy 1.6, and this code does get exercised by some of the tests, as evidenced by the tests failing before I made the two last commits to make the code compatible with Python 2.6.